cpp/src/iterator/iterator.cuh

```diff
@@ -0,0 +1,322 @@
+ * @tparam  T_output  cudf data type of output and `identity` value
+ *                    which is used when null bitmaps flag is false.
+ * @tparam  has_nulls if true, this struct holds only data array.
+ *                    else, this struct holds data array and
+ *                    bitmask array and identity value
+ * -------------------------------------------------------------------------**/
+template <typename T_element, typename T_output, bool has_nulls>
+struct pair_accessor;
+
+template <typename T_element, typename T_output>
+struct pair_accessor<T_element, T_output, true> : public value_accessor<T_element, T_output, true>
+{
+  pair_accessor(T_element const* e, bit_mask::bit_mask_t const* b, T_output i)
+    : value_accessor<T_element, T_output, true>(e, b, i) {};
+
+  CUDA_HOST_DEVICE_CALLABLE
+  thrust::pair<T_output, bool> operator()(gdf_index_type i) const {
+    return bit_mask::is_valid(this->bitmask, i) ?
+        thrust::make_pair(static_cast<T_output>(this->elements[i]), true) :
+        thrust::make_pair(this->identity, false) ;
+  }
+};
+
+template <typename T_element, typename T_output>
+struct pair_accessor<T_element, T_output, false> : public value_accessor<T_element, T_output, false>
+{
+  pair_accessor(T_element const* e, bit_mask::bit_mask_t const* b , T_output i)
+    : value_accessor<T_element, T_output, false>(e, b, i) {};
+
+  CUDA_HOST_DEVICE_CALLABLE
+  thrust::pair<T_output, bool> operator()(gdf_index_type i) const {
+    return thrust::make_pair(static_cast<T_output>(this->elements[i]), true);
+  }
+};
+
+// ---------------------------------------------------------------------------
+// helper functions to make iterator
+
+/** -------------------------------------------------------------------------*
+ * @brief helper function to make a cudf column iterator
+ * Input iterator which can be used for cub and thrust.
+ * The iterator returns same cudf data type of input: `T_element`.
+ *
+ * @tparam has_nulls True if the data has valid bit mask, False else
+ * @tparam T_element The cudf data type of input array
+ * @tparam T_output  cudf data type of output and `identity` value
+ *                   which is used when null bitmaps flag is false.
+ * @tparam Iterator_Index
+ *                   The base iterator which gives the index of array.
+ *                   The default is `thrust::counting_iterator`
+ *
+ * @param[in] data     The pointer of column data array
+ * @param[in] valid    The pointer of null bitmask of column
+ * @param[in] identity The identity value used when the mask value is false
+ * @param[in] it       The index iterator, `thrust::counting_iterator` by default
+ * -------------------------------------------------------------------------**/
+template <bool has_nulls, typename T_element, typename T_output = T_element,
+    typename Iterator_Index=thrust::counting_iterator<gdf_index_type> >
+auto make_iterator(const T_element *data, const bit_mask::bit_mask_t *valid,
```

Comment:
I'd default `valid` and `identity`, because it doesn't make sense to have to provide arguments for those if `has_nulls==false`.  Then add a `if(has_nulls == true)` to ensure that arguments were passed in.

Reply:
@jrhemstad 
>  Then add a if(has_nulls == true) to ensure that arguments were passed in.
How can ensure that `identity` is passed in when if(has_nulls == true)?
For bitmask, the constructor of ` value_accessor<ElementType, ResultType, true>` check it.

Reply:
I've added  default parameter for `valid` and `identity`, then validation check for `data`, `valid` pointer. There is no check for `identity`. See below:
https://github.com/rapidsai/cudf/blob/6d5e3ffe0029ef002a35808f69aa1108331cc5db/cpp/src/iterator/iterator.cuh#L262-L273